crates/http/src/server.rs

```diff
@@ -0,0 +1,282 @@
+        let start_tcp_backend = self.enable_http1 || self.enable_http2;
+
+        #[cfg(feature = "tls-rustls")]
+        if let Some(_rustls_config) = self.rustls_config {
+            #[cfg(not(feature = "http3"))]
+            let enable_http3 = false;
+            #[cfg(feature = "http3")]
+            let enable_http3 = self.enable_http3;
+
+            match (start_tcp_backend, enable_http3) {
+                #[cfg(feature = "http3")]
+                (false, true) => {
+                    use scuffle_context::ContextFutExt;
+
+                    let backend = crate::backend::h3::Http3Backend {
+                        ctx: self.ctx.clone(),
+                        bind: self.bind,
+                    };
+
+                    match backend.run(self.service_factory, _rustls_config).with_context(self.ctx).await {
+                        Some(res) => return res,
+                        None => return Ok(()),
+                    }
+                }
+                #[cfg(any(feature = "http1", feature = "http2"))]
+                (true, false) => {
+                    use scuffle_context::ContextFutExt;
+
+                    let backend = crate::backend::hyper::secure::SecureBackend {
+                        ctx: self.ctx.clone(),
+                        bind: self.bind,
+                        #[cfg(feature = "http1")]
+                        http1_enabled: self.enable_http1,
+                        #[cfg(feature = "http2")]
+                        http2_enabled: self.enable_http2,
+                    };
+
+                    match backend.run(self.service_factory, _rustls_config).with_context(self.ctx).await {
+                        Some(res) => return res,
+                        None => return Ok(()),
+                    }
+                }
+                #[cfg(all(any(feature = "http1", feature = "http2"), feature = "http3"))]
+                (true, true) => {
+                    use scuffle_context::ContextFutExt;
+
+                    let hyper = crate::backend::hyper::secure::SecureBackend {
+                        ctx: self.ctx.clone(),
+                        bind: self.bind,
+                        #[cfg(feature = "http1")]
+                        http1_enabled: self.enable_http1,
+                        #[cfg(feature = "http2")]
+                        http2_enabled: self.enable_http2,
+                    }
+                    .run(self.service_factory.clone(), _rustls_config.clone());
+                    let hyper = std::pin::pin!(hyper);
+
+                    let mut http3 = crate::backend::h3::Http3Backend {
+                        ctx: self.ctx.clone(),
+                        bind: self.bind,
+                    }
+                    .run(self.service_factory, _rustls_config);
+                    let http3 = std::pin::pin!(http3);
+
+                    let res = futures::future::select(hyper, http3).with_context(self.ctx).await;
```

Comment:
we shouldnt use `with_context` here because this will kill the select without a graceful shutdown